akka_microservices/ibex35Microservice/src/main/scala/com/tapiporla/microservices/retrievers/common/ElasticDAO.scala

```diff
@@ -0,0 +1,174 @@
+import com.sksamuel.elastic4s.ElasticsearchClientUri
+import com.sksamuel.elastic4s.indexes.CreateIndexDefinition
+import com.sksamuel.elastic4s.searches.RichSearchResponse
+import com.sksamuel.elastic4s.searches.queries.QueryDefinition
+import com.sksamuel.elastic4s.searches.sort.FieldSortDefinition
+import com.sksamuel.elastic4s.xpack.security.XPackElasticClient
+import com.tapiporla.microservices.retrievers.CausedBy
+import com.tapiporla.microservices.retrievers.common.ElasticDAO._
+import com.tapiporla.microservices.retrievers.common.model.ElasticDocumentInsertable
+import org.elasticsearch.ResourceAlreadyExistsException
+import org.elasticsearch.action.support.WriteRequest.RefreshPolicy
+import org.elasticsearch.common.settings.Settings
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+object ElasticDAO {
+
+  case class SaveInIndex(index: String, typeName: String, jsonDocs: Seq[ElasticDocumentInsertable])
+
+  case class ErrorSavingData(
+                              error: Exception,
+                              index: String,
+                              typeName: String,
+                              jsonDocs: Seq[ElasticDocumentInsertable]
+                            )
+
+  case class DataSavedConfirmation(index: String, typeName: String, jsonDocs: Seq[ElasticDocumentInsertable])
+
+  case class DeleteFromIndex(
+                              index: String,
+                              typeName: String,
+                              where: Option[QueryDefinition] = None
+                            )
+
+  case class ErrorDeletingData(
+                              error: Exception,
+                              index: String,
+                              typeName: String,
+                              originalRQ: DeleteFromIndex
+                              )
+
+  case class DeleteConfirmation(index: String, typeName: String, originalRQ: DeleteFromIndex)
+
+  case class RetrieveAllFromIndexSorted(
+                                         index: String,
+                                         typeName: String,
+                                         where: Option[QueryDefinition] = None,
+                                         sort: Option[FieldSortDefinition] = None,
+                                         limit: Option[Int] = None)
+
+  case class DataRetrieved(
+                            index: String,
+                            typeName: String,
+                            data: RichSearchResponse,
+                            originalRQ: RetrieveAllFromIndexSorted
+                          )
+
+  case class ErrorRetrievingData(
+                                  error: Exception,
+                                  index: String,
+                                  typeName: String,
+                                  originalRQ: RetrieveAllFromIndexSorted
+                                )
+
+  object InitElasticDAO
```

Comment:
I would suggest using case objects if you may want to use these in some pattern matching

Reply:
Done